File: ruby/analyzer.rb
# analyzer.rb — Metrics, health reports, and pattern detection in Ruby
# Showcases: blocks, Proc, lambda, method objects, Comparable, memoization, struct
# frozen_string_literal: true

require_relative 'models'
require 'forwardable'

# ── MetricSnapshot ────────────────────────────────────────────────────────────

MetricSnapshot = Struct.new(
  :total, :by_status, :by_priority,
  :completion_rate, :overdue_count, :avg_age_days, :priority_debt_score,
  keyword_init: true
) do
  def summary
    format(
      'total=%d, done=%.0f%%, overdue=%d, debt=%.2f',
      total, completion_rate * 100, overdue_count, priority_debt_score
    )
  end

  def to_h
    super.merge(completion_pct: (completion_rate * 100).round(1))
  end
end

# ── HealthReport ──────────────────────────────────────────────────────────────

class HealthReport
  THRESHOLDS = {
    critical_overdue: 10,
    high_backlog: 50,
    low_completion: 0.3,
  }.freeze

  attr_reader :snapshot, :warnings, :recommendations

  def initialize(snapshot)
    @snapshot        = snapshot
    @warnings        = []
    @recommendations = []
    evaluate!
  end

  def healthy?
    @warnings.empty?
  end

  def to_s
    return '✅ Healthy' if healthy?

    lines = ['⚠️ Issues detected:']
    @warnings.each { |w| lines << "  - #{w}" }
    lines << 'Recommendations:'
    @recommendations.each { |r| lines << "  → #{r}" }
    lines.join("\n")
  end

  private

  def evaluate!
    s = @snapshot
    if s.overdue_count >= THRESHOLDS[:critical_overdue]
      @warnings        << "#{s.overdue_count} overdue tasks — critical backlog"
      @recommendations << 'Triage overdue tasks immediately'
    end
    if s.completion_rate < THRESHOLDS[:low_completion]
      @warnings        << "Low completion rate: #{(s.completion_rate * 100).round}%"
      @recommendations << 'Break large tasks into smaller subtasks'
    end
    if s.total >= THRESHOLDS[:high_backlog]
      @warnings        << "Large backlog: #{s.total} tasks"
      @recommendations << 'Archive or cancel stale tasks'
    end
  end
end

# ── TaskAnalyzer ──────────────────────────────────────────────────────────────

class TaskAnalyzer
  extend Forwardable

  def_delegators :@collection, :each, :size, :empty?

  def initialize(collection)
    @collection = collection
    # Memoised internal cache
    @_snapshot = nil
  end

  # ── Metrics ───────────────────────────────────────────────────────────

  def snapshot
    @_snapshot ||= compute_snapshot
  end

  def health_report
    HealthReport.new(snapshot)
  end

  def invalidate_cache!
    @_snapshot = nil
  end

  # ── Scoring — demonstrates lambda and proc ────────────────────────────

  DEFAULT_SCORER = lambda do |task|
    base    = Priority.weight(task.priority) * 10.0
    age_f   = Math.log1p(task.age_days) * 2.0
    overdue = task.overdue? ? 20.0 : 0.0
    effort  = task.estimate_effort * 0.5
    base + age_f + overdue - effort
  end

  def rank(top_n: nil, scorer: DEFAULT_SCORER)
    active = @collection.select { |t| Status::ACTIVE.include?(t.status) }
    sorted = active.sort_by { |t| -scorer.call(t) }
    top_n ? sorted.first(top_n) : sorted
  end

  # ── Pattern detection — demonstrates blocks ───────────────────────────

  PATTERNS = {
    stale_todo:            ->(t) { t.status == Status::TODO && t.age_days > 30 },
    high_priority_blocked: ->(t) { t.status == Status::BLOCKED && Priority.urgent?(t.priority) },
    overdue_critical:      ->(t) { t.overdue? && t.priority == Priority::CRITICAL },
    zero_progress_large:   ->(t) { t.subtasks.size > 3 && t.completion_rate.zero? },
  }.freeze

  def detect_patterns
    PATTERNS.transform_values do |pred|
      @collection.select(&pred)
    end
  end

  def stale_tasks(min_age: 30)
    @collection.select { |t| t.status == Status::TODO && t.age_days >= min_age }
  end

  def tag_distribution
    @collection.each_with_object(Hash.new(0)) do |task, counts|
      task.tags.each { |tag| counts[tag.name] += 1 }
    end
  end

  # ── Enumerable-style API ──────────────────────────────────────────────

  def each_critical
    return enum_for(:each_critical) unless block_given?
    @collection.each do |task|
      yield task if task.priority == Priority::CRITICAL && !Status.terminal?(task.status)
    end
  end

  def map_scores(scorer: DEFAULT_SCORER)
    @collection.map { |t| [t, scorer.call(t)] }.to_h
  end

  # ── Private ───────────────────────────────────────────────────────────

  private

  def compute_snapshot
    tasks = @collection.to_a
    return MetricSnapshot.new(total: 0, by_status: {}, by_priority: {},
                               completion_rate: 0.0, overdue_count: 0,
                               avg_age_days: 0.0, priority_debt_score: 0.0) if tasks.empty?

    by_status   = tasks.group_by(&:status).transform_values(&:count)
    by_priority = tasks.group_by(&:priority).transform_values(&:count)
    done        = by_status.fetch(Status::DONE, 0)
    overdue     = tasks.count(&:overdue?)
    avg_age     = tasks.sum(&:age_days).to_f / tasks.size
    debt        = compute_priority_debt(tasks)

    MetricSnapshot.new(
      total:                tasks.size,
      by_status:            by_status,
      by_priority:          by_priority,
      completion_rate:      done.to_f / tasks.size,
      overdue_count:        overdue,
      avg_age_days:         avg_age.round(2),
      priority_debt_score:  debt.round(2),
    )
  end

  def compute_priority_debt(tasks)
    tasks.reduce(0.0) do |sum, task|
      next sum unless Priority.urgent?(task.priority) && !Status.terminal?(task.status)
      sum + Priority.weight(task.priority) * Math.log1p(task.age_days)
    end
  end

  # ── Dead code ──────────────────────────────────────────────────────────

  # Was an early text-only report. Never called in current code.
  def legacy_text_report
    "Total: #{snapshot.total}, Done: #{snapshot.by_status.fetch(Status::DONE, 0)}"
  end
end

File: java/Analyzer.java
package taskmanager;

import java.util.*;
import java.util.stream.*;
import java.util.function.*;

/**
 * Analyzer.java — Computes metrics and reports on a TaskManager.
 * Showcases: exception hierarchy, try-with-resources, functional interfaces,
 * sealed classes (Java 17+).
 */
public class Analyzer {

    // ── Custom exceptions ─────────────────────────────────────────────────────

    public static class AnalysisException extends RuntimeException {
        public AnalysisException(String message) {
            super(message);
        }

        public AnalysisException(String message, Throwable cause) {
            super(message, cause);
        }
    }

    public static class InsufficientDataException extends AnalysisException {
        private final int required;
        private final int found;

        public InsufficientDataException(int required, int found) {
            super(String.format("Need at least %d tasks, found %d", required, found));
            this.required = required;
            this.found = found;
        }

        public int getRequired() {
            return required;
        }

        public int getFound() {
            return found;
        }
    }

    // ── Report types (sealed hierarchy) ──────────────────────────────────────

    public sealed interface AnalysisResult permits Analyzer.OkResult, Analyzer.WarnResult, Analyzer.ErrorResult {
    }

    public record OkResult(MetricSnapshot snapshot) implements AnalysisResult {
    }

    public record WarnResult(MetricSnapshot snapshot, List<String> warnings) implements AnalysisResult {
    }

    public record ErrorResult(String reason, Throwable cause) implements AnalysisResult {
    }

    // ── MetricSnapshot ────────────────────────────────────────────────────────

    public record MetricSnapshot(
            int total,
            Map<String, Long> byStatus,
            Map<String, Long> byPriority,
            double completionRate,
            long overdueCount,
            double avgEffort,
            double priorityDebtScore) {
        public String summary() {
            return String.format(
                    "Total=%d, Done=%.0f%%, Overdue=%d, PriorityDebt=%.2f",
                    total, completionRate * 100, overdueCount, priorityDebtScore);
        }
    }

    // ── Analyzer ──────────────────────────────────────────────────────────────

    private final TaskManager<Task> manager;
    private static final int MIN_TASKS_FOR_ANALYSIS = 1;

    public Analyzer(TaskManager<Task> manager) {
        this.manager = Objects.requireNonNull(manager, "manager must not be null");
    }

    public MetricSnapshot computeSnapshot() {
        List<Task> tasks = manager.all();
        if (tasks.size() < MIN_TASKS_FOR_ANALYSIS) {
            throw new InsufficientDataException(MIN_TASKS_FOR_ANALYSIS, tasks.size());
        }

        Map<String, Long> byStatus = tasks.stream()
                .collect(Collectors.groupingBy(t -> t.getStatus().name(), Collectors.counting()));

        Map<String, Long> byPriority = tasks.stream()
                .collect(Collectors.groupingBy(t -> t.getPriority().name(), Collectors.counting()));

        double completionRate = manager.overallCompletionRate();
        long overdueCount = tasks.stream().filter(Task::isOverdue).count();
        double avgEffort = manager.averageEffort().orElse(0.0);
        double priorityDebt = computePriorityDebt(tasks);

        return new MetricSnapshot(tasks.size(), byStatus, byPriority,
                completionRate, overdueCount, avgEffort, priorityDebt);
    }

    public AnalysisResult analyse() {
        try {
            MetricSnapshot snapshot = computeSnapshot();
            List<String> warnings = collectWarnings(snapshot);
            if (warnings.isEmpty()) {
                return new OkResult(snapshot);
            } else {
                return new WarnResult(snapshot, warnings);
            }
        } catch (InsufficientDataException e) {
            return new ErrorResult("Not enough data: " + e.getMessage(), e);
        } catch (Exception e) {
            return new ErrorResult("Unexpected error during analysis", e);
        }
    }

    public List<Task> rankByUrgency(int topN) {
        ToDoubleFunction<Task> scorer = t -> {
            double base = t.getPriority().getWeight() * 10.0;
            double age = Math.log1p(t.getAgeDays()) * 2.0;
            double overdue = t.isOverdue() ? 20.0 : 0.0;
            double effort = t.estimateEffort() * 0.5;
            return base + age + overdue - effort;
        };
        return manager.topN(topN, scorer);
    }

    public Map<String, List<Task>> detectPatterns() {
        Map<String, List<Task>> patterns = new LinkedHashMap<>();

        // Stale: TODO for > 30 days
        patterns.put("stale_todo", manager.search(
                t -> t.getStatus() == Task.Status.TODO && t.getAgeDays() > 30));

        // High-priority blocked
        patterns.put("high_priority_blocked", manager.search(
                t -> t.getStatus() == Task.Status.BLOCKED && t.getPriority().isUrgentOrAbove()));

        // Overdue high priority
        patterns.put("overdue_high_priority", manager.search(
                t -> t.isOverdue() && t.getPriority().isUrgentOrAbove()));

        return patterns;
    }

    // ── Private helpers ───────────────────────────────────────────────────────

    private List<String> collectWarnings(MetricSnapshot snapshot) {
        List<String> warnings = new ArrayList<>();
        if (snapshot.overdueCount() >= 10)
            warnings.add("Critical: " + snapshot.overdueCount() + " overdue tasks");
        if (snapshot.completionRate() < 0.3)
            warnings.add("Low completion rate: " + String.format("%.0f%%", snapshot.completionRate() * 100));
        if (snapshot.total() > 50)
            warnings.add("Large backlog: " + snapshot.total() + " tasks");
        if (snapshot.priorityDebtScore() > 100)
            warnings.add("High priority debt: " + String.format("%.1f", snapshot.priorityDebtScore()));
        return warnings;
    }

    private double computePriorityDebt(List<Task> tasks) {
        return tasks.stream()
                .filter(t -> t.getPriority().isUrgentOrAbove() && !t.getStatus().isTerminal())
                .mapToDouble(t -> t.getPriority().getWeight() * Math.log1p(t.getAgeDays()))
                .sum();
    }

    // ── Dead code ─────────────────────────────────────────────────────────────

    /**
     * @deprecated Replaced by computeSnapshot(). Never called in the current
     *             codebase.
     */
    @Deprecated
    private double legacyScore(Task t) {
        return t.getPriority().getWeight() * 5.0;
    }
}
